File: Práctica4/lib/SpecificDamage.rb
#encoding:utf-8

require_relative 'Damage'
require_relative 'SpecificDamageToUI'

module Deepspace
  # Clase que representa un daño específico
  class SpecificDamage < Damage
    public_class_method :new

    # Constructor
    # @param _nShields [Integer] numero de escudos que se pierden
    # @param _weapons [Array<WeaponType>] array de armas que se pierden
    def initialize(_weapons, _nShields)
      super(_nShields)

      # @!attribute [Array<WeaponType>] array de armas que se pierden
      @weapons = []
    end

    # Getters
    attr_reader :weapons

    # Comprube si el daño hace efecto
    # @return [Boolean] true en caso afirmativo, false en caso contrario
    # --Overriden
    def hasNoEffect
      (super && @weapons.length == 0)
    end

    # Busca en un vector de armas dados la primera coincidencia con el tipo de arma dado
    # @param w [Array<Weapon>] el array de armas
    # @param t [WeaponType] el tipo de arma
    # @return [Integer] la posicion del arma. Si no se encuentra devuelve -1
    def arrayContainsType(w, t)
      i = 0
      w.each do |weapon_aux|
        if weapon_aux.type == t
          return i
        else
          i += 1
        end
      end

      # No element found
      return -1
    end

    # Crea un parametro ajustado a las instancias pasadas como argumentos
    # @param w [Array<Weapon>] el array de armas a ajustar
    # @param s [Array<ShieldBooster>] el array de escudos a ajustar
    # @return [SpecificDamage] la copia ajustada
    # --Overriden
    def adjust(w, s)
      weapons_copy = @weapons.clone

      new_weapons = w.map do |weapon|
        weapons_copy.delete_at(weapons_copy.index(weapon.type) || weapons_copy.length)
      end

      new_weapons.compact!

      self.class.new(new_weapons, super(s))
    end

    # Devuelve una copia de la instancia actual
    # @return [SpecificDamage] copia
    def copy
      self.class.new(weapons, nShields)
    end

    # Elimina un arma
    # Busca la primera ocurrencia de tipo de arma y elimina dicha pisicion
    # @param w [Weapon] el arma a eliminar
    def discardWeapon(w)
      if @weapons.length != 0
        position = @weapons.index(w.type)
        if position != nil
          @weapons.delete_at(position)
        else
          raise "WARNING! No weapon type match at Damage.discardWeapon()"
        end
      end
    end

    # String representation of the object
    # @return [String] string representation
    def to_s
      message = "[Specific Damage] -> Shields: #{@nShields}, Weapon types: " + getWeaponInfo
      return message
    end

    # To UI
    def getUIversion
      return SpecificDamageToUI.new(self)
    end
  end
end

File: Práctica5/java/src/deepspace/SpecificDamage.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package deepspace;

import java.util.ArrayList;
import java.util.Iterator;

/**
 * SpecificDamage está compuesto de:
 * - número de escudos a perder
 * - lista de armas a perder
 */
public class SpecificDamage extends Damage {
    /**
     * Array con las armas a perder
     */
    ArrayList<WeaponType> weapons;
    
    /**
     * Constructor
     * @param _weapons array con los tipos de armas a perder
     * @param _nShields número de escudos a perder
     */
    public SpecificDamage(ArrayList<WeaponType> _weapons, int _nShields){
        super(_nShields);
        if (_weapons != null)
            weapons = new ArrayList<>(_weapons);
        else
            weapons = new ArrayList<>();
    }
    
    /**
     * Constructor de copia
     * @param other instancia a copiar
     */
    public SpecificDamage(SpecificDamage other){
        this(other.weapons, other.getNShields());
    }
    
    // Getters
    
    /**
     * Copy getter
     * @return copia de la instancia
     */
    @Override
    public SpecificDamage copy(){
        return new SpecificDamage(this);
    }
    
    /**
     * Getter de weapons
     * @return copy de weapons
     */
    public ArrayList<WeaponType> getWeapons(){
        if (weapons != null)
            return new ArrayList<>(weapons);
        else
            return new ArrayList<>();
    }
    
    /**
     * Comprueba si el daño tiene efecto o no
     * @return true en caso de no tener efecto, false en caso contrario
     */
    @Override
    public boolean hasNoEffect(){
        if (weapons != null)
            return weapons.isEmpty() && getNShields() == 0;
        else
            return getNShields() == 0;
    }
    
    
    // Setters
    
    /**
     * Busca el primer elemento de un tipo de arma en un array dado
     * @param w lista de armas
     * @param t tipo de arma
     * @return devuelve la posición. Si no se encuentra, devuelve -1
     */
    private int arrayContainsType(ArrayList<Weapon> w, WeaponType t){
        Iterator<Weapon> it = w.iterator();
        int i=0;
        while(it.hasNext()){
            if (t == it.next().getType()){
                return i;
            }
            i++;
        }
        return -1;
    }
    
    /**
     * Crea un objeto ajustado 
     * @param w armas a ajustar
     * @param s escudos a ajustar
     * @return copia del objecto ajustado
     */
    @Override
    public SpecificDamage adjust(ArrayList<Weapon> w, ArrayList<ShieldBooster> s){
        int shields = Integer.min(s.size(), getNShields());
        ArrayList<Weapon> wAux = new ArrayList<>(w);
       ArrayList<WeaponType> toSet = new ArrayList<>();
       
       for(WeaponType element: weapons) {
           int index = arrayContainsType(wAux, element);
          
          if(index != -1){  //If the element is found
              toSet.add(element); //It can be removed
              wAux.remove(index);
          }
       }
        
        return new SpecificDamage(toSet, shields);
    }
    
    /**
     * Elimina un arma dada
     * @param w arma a descartar
     * */
    @Override
    public void discardWeapon(Weapon w){
        if (!weapons.isEmpty()){
                weapons.remove(w.getType());
        }
    }
    
    /**
     * Get the string representation of the object
     * @return the string representation
     **/
    @Override
    public String toString(){
        return  "Damage(" +
                "weapons = " + weapons +
                ", nShields = " + getNShields() +
                ")";
    }

    /**
     * Gets the UI version of the object
     *
     * @return the UI version of the object
     * */
    @Override
    public SpecificDamageToUI getUIversion(){
        return new SpecificDamageToUI(this);
    }
}
